Planout ES6 Port

rawls238 opened the pull request:
@eytan 

This ports the Planout python library almost verbatim to Javascript (ES6).

There are still two major TODOs before this can be merged in:
1) Right now I use 13 bits of precision for the getHash function which isn't consistent with the other planout libraries. This leads to different assignments for the JS library vs the Python library for the same salt. We can't use the basic JS Number type for this, it's going to require third-party library. I've tried a couple and none has worked correctly, but I will try a few more. 

2) Get the compilation to ES5 working properly so that people can also use Planout in their ES5 code as well.

rawls238 commented:
https://www.npmjs.com/package/bignumber.js seemed to do the trick.

Only 2) is left as a significant TODO now

rawls238 commented:
2) is kind of done now, but the minified JS file is 1.9 MB which is insane. Going to try to figure out how I can reduce the size of it - I think that the inflated size is partially due to the libraries that I'm including but haven't confirmed that yet.

rawls238 commented on js/es6/interpreter.js:
```diff
@@ -0,0 +1,94 @@
+	}
+
+	set_env(new_env) {
+		this._env = $.extend(true, {}, new_env);
```
this function doesn't appear to be used anywhere - might just remove it or implement some naive version of deep copy here instead of relying on jQuery for this. As of right now this is the only place where we use jQuery for anything.

gusvargas commented on js/__tests__/testAssignment.js:
```diff
@@ -0,0 +1,33 @@
+var UniformChoice = require('../es6/ops/random').UniformChoice;
+
+var tester_unit = '4';
+var tester_salt = 'test_salt';
```
These could be `const`

gusvargas commented:
A couple common stylistic things to fix:

- Indentation is weird in a lot of places
- Use camelCase instead of snake_case

gusvargas commented on js/es6/ops/base.js:
```diff
@@ -0,0 +1,118 @@
+
+	getArgMixed(name) {
+		if (this.args[name] === undefined) {
+			throw ("Missing argument " + name);
```
No need for the parens here. You can use template strings also:

```javascript
throw `Missing argument ${name}`
```

gusvargas commented on js/es6/ops/base.js:
```diff
@@ -0,0 +1,118 @@
+		var self = this;
+		Object.keys(this.args).forEach(function (key) {
+  			self.args[key] = mapper.evaluate(self.args[key]);
+		});
```
If you use [arrow functions](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Arrow_functions) it will preserve `this`.

eytan commented:
Yeah, 1.9MB is pretty unreasonable... underscore makes code look prettier, but it seems like it could be adding a lot of unnecessary weight.  Perhaps we could ditch it in favor of vanilla JS?

eytan commented on js/es6/experiment.js:
```diff
@@ -0,0 +1,163 @@
+			'inputs': this.inputs,
+			'params': this._assignment.get_params()
+		};
+		_.extend(d, extras);
```
Can you use `Object.assign()` here?

eytan commented on js/es6/interpreter.js:
```diff
@@ -0,0 +1,95 @@
+	}
+
+	evaluate(planout_code) {
+		if (Object.prototype.toString.call( planout_code ) === '[object Object]' && planout_code.op) {
```
Is there a more straightforward way to do the introspection here?

eytan commented on js/es6/ops/core.js:
```diff
@@ -0,0 +1,236 @@
+
+class Seq extends PlanOutOp {
+	execute(mapper) {
+		_.each(this.getArgList('seq'), function(op) {
```
can you do something like
```javascript
this.getArgList('seq').forEach(op => mapper.evaluate(op));
```
here?

eytan commented:
This is *awesome*!  I can't tell you how excited I am to see a port of PlanOut to JavaScript!

A few last things.  Can you please add:
- [ ] A brief README.md that describes what parts of PlanOut have been implemented and what is missing, and an example of how you would use the package (either using the native API or the interpreter).  It looks like this includes basically everything besides namespaces, and the setter sugar for the native API.
- [ ] Some (informal) verification that the PlanOut script, when executed using this JS implementation, produces the same results as the Python implementation.  This could just be a gist you put somewhere.
- [ ] Move the files from the top-level directory `js` to `alpha/js`, since this implementation is still pretty hot.

It would also be cool to see a demo node.js app that uses the package at some point.

rawls238 commented:
@eytan I'll definitely add all those things when I get the final implementation sorted out and before I merge this in.

rawls238 commented:
@eytan I got the minified file size down to 50 KB...there was a bug with my gruntfile.js that was causing the file size to blow up. I think that everything should be ready to go here now.

With regards to the verification of the equivalence of the Python and JS implementations, I think the unit tests should suffice. The unit tests I use here are identical in almost every respect to those used in the python code and given that all the tests pass I think I can say with some reasonable degree of certainty that they should be aligned. Let me know if you think some additional verification would be necessary.

I'll write up the README for this project now and I'll start on a small sample app. Let me know what else you think should be done before we can merge this in

rawls238 commented:
also when this is merged in I'll be making it available on npm.

rawls238 commented:
@eytan this should be all set to merge so let me know if there are any additional concerns. Let me know if you want this project to be in its own repo or something (similar to the PHP port), but I'm planning on using it later this week for a project so I'll probably publish it to npm on Wednesday or Thursday if that's okay with you.

eytan commented:
This is great, Guy! I am excited to merge it into the main PlanOut repository --- I just want to verify one thing --- does this produce random assignments that are consistent with the Python version of PlanOu?.  For example, if one ran the demo scripts from the Python implementation using JS, would it produce the exact same output?  This is key.  It would be great if you could informally verify this, and super bonus if you could run the JSON scripts from the Python demo and show that the outputs are the same.

One last thing, I don't understand the purpose of:
```
args.val++; //args is an arbitrary object passed in on initialization
```
in the README. What is this trying to illustrate?  Nonetheless, it seems like a bit beyond the scope of what one typically highlights in the README -- you just want to focus on how to get started, and ideally show that it's pretty easy -- not get into more advanced aspects the semantics.

Things are a little crazy right now for me, but I would love to highlight this implementation prominently on the PlanOut website, and include some more in-depth documentation in the docs.

rawls238 commented:
@eytan you're right about the README, I'm not sure why I added that line.

I added a couple of experiments from the demos folder and verified that the outputs are identical: 

Python run: 

Experiment 1: <code> [(10, None), (1, 64), (10, 320), (1, 16), (10, 320), (1, 64), (1, None), (1, None), (1, 32), (10, None)]
</code>

Experiment 3:
<code>
{'inputs': {'userid': 0}, 'name': 'Exp3', 'checksum': '3d32c926', 'params': {'has_feed_stories': 1, 'button_text': "I'm a voter", 'has_banner': 1}, 'time': 1431399039, 'salt': 'Exp3'}
{'inputs': {'userid': 1}, 'name': 'Exp3', 'checksum': '3d32c926', 'params': {'has_feed_stories': 1, 'button_text': "I'm voting", 'has_banner': 1}, 'time': 1431399039, 'salt': 'Exp3'}
{'inputs': {'userid': 2}, 'name': 'Exp3', 'checksum': '3d32c926', 'params': {'has_feed_stories': 1, 'button_text': "I'm voting", 'has_banner': 1}, 'time': 1431399039, 'salt': 'Exp3'}
{'inputs': {'userid': 3}, 'name': 'Exp3', 'checksum': '3d32c926', 'params': {'has_feed_stories': 1, 'button_text': "I'm a voter", 'has_banner': 1}, 'time': 1431399039, 'salt': 'Exp3'}
{'inputs': {'userid': 4}, 'name': 'Exp3', 'checksum': '3d32c926', 'params': {'has_feed_stories': 1, 'button_text': "I'm voting", 'has_banner': 1}, 'time': 1431399039, 'salt': 'Exp3'}
</code>

Javascript run: 
<code>
Experiment 1 results 10,,1,64,10,320,1,16,10,320,1,64,1,,1,,1,32,10,
Experiment 3 results I'm a voter,I'm voting,I'm voting,I'm a voter,I'm voting
</code>

rawls238 commented:
It would be awesome to add some more in-depth documentation on the website, also would be nice to have a nicer way to interact with this in ES5 (perhaps using a pattern like this: https://gist.github.com/juandopazo/1367191) since the current way I'm doing in the example scripts is a bit ugly.
